File: python/src/dflow/utils/pagination.py
"""Pagination utilities for DFlow API responses."""

from collections.abc import AsyncGenerator, Callable, Generator
from typing import Any, TypeVar

T = TypeVar("T")
TResponse = TypeVar("TResponse")


def paginate(
    fetch_page: Callable[[dict[str, Any]], TResponse],
    get_items: Callable[[TResponse], list[T]],
    get_cursor: Callable[[TResponse], str | None] | None = None,
    max_items: int | None = None,
    page_size: int | None = None,
) -> Generator[T, None, None]:
    """Create a generator that automatically paginates through all results.

    Example:
        >>> from dflow.utils import paginate
        >>>
        >>> # Iterate through all markets
        >>> for market in paginate(
        ...     lambda params: dflow.markets.get_markets(**params),
        ...     get_items=lambda r: r.markets,
        ... ):
        ...     print(market.ticker, market.yes_price)
        >>>
        >>> # With a limit
        >>> for event in paginate(
        ...     lambda params: dflow.events.get_events(**params),
        ...     get_items=lambda r: r.events,
        ...     max_items=100,
        ... ):
        ...     print(event.title)

    Args:
        fetch_page: Function that fetches a page given pagination params
        get_items: Function to extract items array from response
        get_cursor: Function to extract cursor from response (default: r.cursor)
        max_items: Maximum number of items to fetch in total (default: unlimited)
        page_size: Number of items per page (default: API default)

    Yields:
        Individual items from each page
    """
    if get_cursor is None:

        def get_cursor(r: Any) -> str | None:
            return getattr(r, "cursor", None)

    cursor: str | None = None
    items_yielded = 0

    while True:
        params: dict[str, Any] = {}
        if cursor:
            params["cursor"] = cursor
        if page_size:
            params["limit"] = page_size

        response = fetch_page(params)
        items = get_items(response)

        for item in items:
            yield item
            items_yielded += 1

            if max_items is not None and items_yielded >= max_items:
                return

        cursor = get_cursor(response)
        if not cursor:
            break


def collect_all(
    fetch_page: Callable[[dict[str, Any]], TResponse],
    get_items: Callable[[TResponse], list[T]],
    get_cursor: Callable[[TResponse], str | None] | None = None,
    max_items: int | None = None,
    page_size: int | None = None,
) -> list[T]:
    """Collect all items from a paginated endpoint into a list.

    Example:
        >>> from dflow.utils import collect_all
        >>>
        >>> # Get all markets as a list
        >>> all_markets = collect_all(
        ...     lambda params: dflow.markets.get_markets(**params),
        ...     get_items=lambda r: r.markets,
        ... )
        >>> print(f"Found {len(all_markets)} markets")

    Args:
        fetch_page: Function that fetches a page given pagination params
        get_items: Function to extract items array from response
        get_cursor: Function to extract cursor from response
        max_items: Maximum number of items to fetch in total
        page_size: Number of items per page

    Returns:
        List of all items
    """
    return list(
        paginate(
            fetch_page=fetch_page,
            get_items=get_items,
            get_cursor=get_cursor,
            max_items=max_items,
            page_size=page_size,
        )
    )


def count_all(
    fetch_page: Callable[[dict[str, Any]], TResponse],
    get_items: Callable[[TResponse], list[T]],
    get_cursor: Callable[[TResponse], str | None] | None = None,
) -> int:
    """Count total items from a paginated endpoint without storing them.

    Example:
        >>> from dflow.utils import count_all
        >>>
        >>> total = count_all(
        ...     lambda params: dflow.markets.get_markets(**params),
        ...     get_items=lambda r: r.markets,
        ... )
        >>> print(f"Total markets: {total}")

    Args:
        fetch_page: Function that fetches a page given pagination params
        get_items: Function to extract items array from response
        get_cursor: Function to extract cursor from response

    Returns:
        Total count of items
    """
    count = 0
    for _ in paginate(
        fetch_page=fetch_page,
        get_items=get_items,
        get_cursor=get_cursor,
    ):
        count += 1
    return count


def find_first(
    fetch_page: Callable[[dict[str, Any]], TResponse],
    get_items: Callable[[TResponse], list[T]],
    predicate: Callable[[T], bool],
    get_cursor: Callable[[TResponse], str | None] | None = None,
) -> T | None:
    """Find the first item matching a predicate from a paginated endpoint.

    Example:
        >>> from dflow.utils import find_first
        >>>
        >>> # Find a specific market
        >>> market = find_first(
        ...     lambda params: dflow.markets.get_markets(**params),
        ...     get_items=lambda r: r.markets,
        ...     predicate=lambda m: "Bitcoin" in m.title,
        ... )
        >>> if market:
        ...     print(f"Found: {market.ticker}")

    Args:
        fetch_page: Function that fetches a page given pagination params
        get_items: Function to extract items array from response
        predicate: Function to test each item
        get_cursor: Function to extract cursor from response

    Returns:
        The first matching item, or None if not found
    """
    for item in paginate(
        fetch_page=fetch_page,
        get_items=get_items,
        get_cursor=get_cursor,
    ):
        if predicate(item):
            return item
    return None


async def paginate_async(
    fetch_page: Callable[[dict[str, Any]], Any],
    get_items: Callable[[TResponse], list[T]],
    get_cursor: Callable[[TResponse], str | None] | None = None,
    max_items: int | None = None,
    page_size: int | None = None,
) -> AsyncGenerator[T, None]:
    """Async version of paginate.

    Example:
        >>> from dflow.utils import paginate_async
        >>>
        >>> async for market in paginate_async(
        ...     lambda params: async_client.markets.get_markets(**params),
        ...     get_items=lambda r: r.markets,
        ... ):
        ...     print(market.ticker)

    Args:
        fetch_page: Async function that fetches a page given pagination params
        get_items: Function to extract items array from response
        get_cursor: Function to extract cursor from response
        max_items: Maximum number of items to fetch in total
        page_size: Number of items per page

    Yields:
        Individual items from each page
    """
    if get_cursor is None:

        def get_cursor(r: Any) -> str | None:
            return getattr(r, "cursor", None)

    cursor: str | None = None
    items_yielded = 0

    while True:
        params: dict[str, Any] = {}
        if cursor:
            params["cursor"] = cursor
        if page_size:
            params["limit"] = page_size

        response = await fetch_page(params)
        items = get_items(response)

        for item in items:
            yield item
            items_yielded += 1

            if max_items is not None and items_yielded >= max_items:
                return

        cursor = get_cursor(response)
        if not cursor:
            break

File: packages/typescript/src/utils/pagination.ts
import type { PaginationParams } from '../types/common.js';

/**
 * Generic paginated response type that works with various API responses.
 * Supports both `data` field and custom field names like `markets`, `events`, etc.
 */
export interface PaginatedResult<T> {
  cursor?: string;
  items: T[];
}

export interface PaginateOptions<TResponse, TItem> {
  /** Maximum number of items to fetch in total (default: unlimited) */
  maxItems?: number;
  /** Number of items per page (default: API default, usually 50) */
  pageSize?: number;
  /**
   * Function to extract items array from response.
   * Required because different APIs use different field names (markets, events, data, etc.)
   */
  getItems: (response: TResponse) => TItem[];
  /** Function to extract cursor from response (default: response.cursor) */
  getCursor?: (response: TResponse) => string | undefined;
}

/**
 * Create an async iterator that automatically paginates through all results.
 *
 * @example
 * ```typescript
 * import { paginate } from 'dflow-sdk';
 *
 * // Iterate through all markets
 * for await (const market of paginate(
 *   (params) => dflow.markets.getMarkets(params),
 *   { getItems: (r) => r.markets }
 * )) {
 *   console.log(market.ticker, market.yesPrice);
 * }
 *
 * // Iterate through all events
 * for await (const event of paginate(
 *   (params) => dflow.events.getEvents({ ...params, status: 'active' }),
 *   { getItems: (r) => r.events, maxItems: 100 }
 * )) {
 *   console.log(event.title);
 * }
 * ```
 *
 * @param fetchPage - Function that fetches a page given pagination params
 * @param options - Pagination options including item extractor
 * @yields Individual items from each page
 */
export async function* paginate<TResponse extends { cursor?: string }, TItem>(
  fetchPage: (params: PaginationParams) => Promise<TResponse>,
  options: PaginateOptions<TResponse, TItem>
): AsyncGenerator<TItem, void, undefined> {
  const { maxItems, pageSize, getItems, getCursor = (r) => r.cursor } = options;

  let cursor: string | undefined;
  let itemsYielded = 0;

  do {
    const response = await fetchPage({
      cursor,
      limit: pageSize,
    });

    const items = getItems(response);
    for (const item of items) {
      yield item;
      itemsYielded++;

      if (maxItems !== undefined && itemsYielded >= maxItems) {
        return;
      }
    }

    cursor = getCursor(response);
  } while (cursor);
}

/**
 * Collect all items from a paginated endpoint into an array.
 *
 * @example
 * ```typescript
 * import { collectAll } from 'dflow-sdk';
 *
 * // Get all markets as an array
 * const allMarkets = await collectAll(
 *   (params) => dflow.markets.getMarkets(params),
 *   { getItems: (r) => r.markets }
 * );
 *
 * console.log(`Found ${allMarkets.length} markets`);
 *
 * // With a limit
 * const first100 = await collectAll(
 *   (params) => dflow.events.getEvents(params),
 *   { getItems: (r) => r.events, maxItems: 100 }
 * );
 * ```
 *
 * @param fetchPage - Function that fetches a page given pagination params
 * @param options - Pagination options including item extractor
 * @returns Array of all items
 */
export async function collectAll<TResponse extends { cursor?: string }, TItem>(
  fetchPage: (params: PaginationParams) => Promise<TResponse>,
  options: PaginateOptions<TResponse, TItem>
): Promise<TItem[]> {
  const items: TItem[] = [];

  for await (const item of paginate(fetchPage, options)) {
    items.push(item);
  }

  return items;
}

/**
 * Count total items from a paginated endpoint without storing them.
 *
 * @example
 * ```typescript
 * import { countAll } from 'dflow-sdk';
 *
 * const totalMarkets = await countAll(
 *   (params) => dflow.markets.getMarkets(params),
 *   { getItems: (r) => r.markets }
 * );
 *
 * console.log(`Total markets: ${totalMarkets}`);
 * ```
 *
 * @param fetchPage - Function that fetches a page given pagination params
 * @param options - Pagination options including item extractor
 * @returns Total count of items
 */
export async function countAll<TResponse extends { cursor?: string }, TItem>(
  fetchPage: (params: PaginationParams) => Promise<TResponse>,
  options: PaginateOptions<TResponse, TItem>
): Promise<number> {
  let count = 0;

  for await (const _ of paginate(fetchPage, options)) {
    count++;
  }

  return count;
}

/**
 * Find the first item matching a predicate from a paginated endpoint.
 *
 * @example
 * ```typescript
 * import { findFirst } from 'dflow-sdk';
 *
 * // Find a specific market by title
 * const market = await findFirst(
 *   (params) => dflow.markets.getMarkets(params),
 *   { getItems: (r) => r.markets },
 *   (m) => m.title.includes('Bitcoin')
 * );
 *
 * if (market) {
 *   console.log('Found:', market.ticker);
 * }
 * ```
 *
 * @param fetchPage - Function that fetches a page given pagination params
 * @param options - Pagination options including item extractor
 * @param predicate - Function to test each item
 * @returns The first matching item, or undefined if not found
 */
export async function findFirst<TResponse extends { cursor?: string }, TItem>(
  fetchPage: (params: PaginationParams) => Promise<TResponse>,
  options: PaginateOptions<TResponse, TItem>,
  predicate: (item: TItem) => boolean
): Promise<TItem | undefined> {
  for await (const item of paginate(fetchPage, options)) {
    if (predicate(item)) {
      return item;
    }
  }

  return undefined;
}
